scrape/scrape.go

```diff
@@ -1835,7 +1835,9 @@ loop:
 	if err == nil {
 		sl.cache.forEachStale(func(lset labels.Labels) bool {
 			// Series no longer exposed, mark it stale.
+			app.SetHints(&storage.AppendHints{DiscardOutOfOrder: true})
 			_, err = app.Append(0, lset, defTime, math.Float64frombits(value.StaleNaN))
+			app.SetHints(nil)
```

Comment:
[The original proposal](https://github.com/prometheus/prometheus/issues/11765) for this was adding hints to the `Append()` method: why did we chose to do a mutable appender with a state instead?

Reply:
Two different implementations were attempted:
- https://github.com/prometheus/prometheus/pull/15184
- https://github.com/prometheus/prometheus/pull/14653

But both were discarded because of too many lines diff without an actual behavior modification being added. We'll move forward with the current approach to move forward. 

I'll probably rename SetHints to SetOptions before merging this PR.